packages/next-safe-action/src/index.ts

```diff
@@ -65,30 +72,48 @@ export const createSafeActionClient = <Context>(createOpts?: SafeClientOpts<Cont
 	// It expects an input schema and a `serverCode` function, so the action
 	// knows what to do on the server when called by the client.
 	// It returns a function callable by the client.
-	const actionBuilder = <const S extends Schema, const Data>(
-		schema: S,
-		serverCode: ServerCodeFn<S, Data, Context>
+	const actionBuilder = <const S extends Schema[], const Data>(
+		...args: [...S, ServerCodeFn<S, Data, Context>]
 	): SafeAction<S, Data> => {
 		// This is the function called by client. If `input` fails the schema
 		// parsing, the function will return a `validationError` object, containing
 		// all the invalid fields provided.
-		return async (clientInput) => {
-			try {
-				const parsedInput = await wrap(schema).validate(clientInput);
 
-				// If schema validation fails.
-				if ("issues" in parsedInput) {
-					return {
-						validationErrors: buildValidationErrors(parsedInput.issues),
-					};
+		const schemas = args.slice(0, -1) as S;
+		const serverCode = args[args.length - 1] as ServerCodeFn<S, Data, Context>;
+
+		return async (...inputs) => {
+			try {
+				const parsedInputs = await Promise.all(
+					inputs.map(async (input, index) => {
+						const schema = schemas[index]!;
+						return await wrap(schema).validate(input);
+					})
+				);
+
+				const failedInputs = parsedInputs.reduce(
+					(prev, parsedInput) => {
+						return "issues" in parsedInput
+							? { ...prev, ...buildValidationErrors(parsedInput.issues) }
+							: prev;
+					},
+					{} as Partial<Record<keyof InferArray<S>[number] | "_root", string[]>>
```

Comment:
I simply merge all validation errors together, but this is a naive approach. Different arguments could have fields that overlap and it is hard to see from which argument a validation error came from.

**Option 1**
We could simply _only_ return the validation errors of the first input that failed and avoid breaking changes.

**Option 2**
Other (preferable, but breaking) option would be to do something like:
```typescript
type ValidationErrors<S extends Schema> =  Partial<Record<keyof Infer<S> | "_root", string[]>>
type SafeActionResult<S extends Schema[], Data> = {
	data?: Data;
	serverError?: string;
	validationErrors: { [K in keyof S]: ValidationErrors<S[K]> }
}
```

So that the user can access validation errors like:
```typescript
const input1Errors = result.validationErrors[0] // full type safety here
const input2Errors = result.validationErrors[1]
```

Reply:
what are your thoughts on this?